juniper_warp/src/lib.rs

```diff
@@ -0,0 +1,363 @@
+        use juniper::tests::model::Database;
+        use juniper::{EmptyMutation, RootNode};
+
+        type Schema = juniper::RootNode<'static, Database, EmptyMutation<Database>>;
+
+        let schema: Schema = RootNode::new(Database::new(), EmptyMutation::<Database>::new());
+
+        let state = warp::any().map(move || Database::new());
+        let filter = warp::path("graphql2").and(make_graphql_filter(schema, state.boxed()));
+
+        let response = request()
+            .method("POST")
+            .path("/graphql2")
+            .header("accept", "application/json")
+            .header("content-type", "application/json")
+            .body(r##"{ "variables": null, "query": "{ hero(episode: NEW_HOPE) { name } }" }"##)
+            .reply(&filter);
+
+        assert_eq!(response.status(), http::StatusCode::OK);
+        assert_eq!(
+            response.headers().get("content-type").unwrap(),
+            "application/json",
+        );
+        assert_eq!(
+            String::from_utf8(response.body().to_vec()).unwrap(),
+            r#"{"data":{"hero":{"name":"R2-D2"}}}"#
+        );
+    }
+}
+
+#[cfg(test)]
+mod tests_http_harness {
+    use super::*;
+    use juniper::http::tests::{run_http_test_suite, HTTPIntegration, TestResponse};
+    use juniper::tests::model::Database;
+    use juniper::EmptyMutation;
+    use juniper::RootNode;
+    use warp;
+    use warp::Filter;
+
+    type Schema = juniper::RootNode<'static, Database, EmptyMutation<Database>>;
+
+    fn warp_server() -> warp::filters::BoxedFilter<(warp::http::Response<Vec<u8>>,)> {
+        let schema: Schema = RootNode::new(Database::new(), EmptyMutation::<Database>::new());
+
+        let state = warp::any().map(move || Database::new());
+        let filter = warp::filters::path::index().and(make_graphql_filter(schema, state.boxed()));
+
+        filter.boxed()
+    }
+
+    struct TestWarpIntegration {
+        filter: warp::filters::BoxedFilter<(warp::http::Response<Vec<u8>>,)>,
+    }
+
+    // This can't be implemented with the From trait since TestResponse is not defined in this crate.
+    fn test_response_from_http_response(response: warp::http::Response<Vec<u8>>) -> TestResponse {
+        TestResponse {
+            status_code: response.status().as_u16() as i32,
+            body: Some(String::from_utf8(response.body().to_owned()).unwrap()),
+            content_type: response
+                .headers()
+                .get("content-type")
+                .expect("missing content-type header in warp response")
+                .to_str()
+                .expect("invalid content-type string")
+                .to_owned(),
+        }
+    }
+
+    impl HTTPIntegration for TestWarpIntegration {
+        fn get(&self, url: &str) -> TestResponse {
+            use percent_encoding::{percent_encode, DEFAULT_ENCODE_SET};
+            let url: String = percent_encode(url.replace("/?", "").as_bytes(), DEFAULT_ENCODE_SET)
```

Comment:
This is fine for now, but I think we'll have the harness send valid encoded URLs so all these integrations don't have to deal with it themselves.

Reply:
Oh, I wasn't clear...I was just musing out loud! I don't think there is a bug in warp or you changes here, but it is actually an issue with the integration test harness. See https://github.com/graphql-rust/juniper/pull/230#issuecomment-414464667 for the same thing with Hyper, and I had to fix it hackily in the iron integration a bit ago in https://github.com/graphql-rust/juniper/pull/208.

Reply:
Oh alright, sorry for the misunderstanding :)